interp/src/primitives/stateful.rs

```diff
@@ -1385,3 +1386,222 @@ impl Primitive for StdMemD4 {
         )
     }
 }
+
+#[derive(Default)]
+pub struct StdFpMultPipe<const SIGNED: bool> {
+    pub width: u64,
+    pub int_width: u64,
+    pub frac_width: u64,
+    pub product: Value,
+    update: Option<Value>,
+    queue: VecDeque<Option<Value>>, //invariant: always length 2.
+}
+
+impl<const SIGNED: bool> StdFpMultPipe<SIGNED> {
+    pub fn from_constants(width: u64, int_width: u64, frac_width: u64) -> Self {
+        assert_eq!(width, int_width + frac_width);
+        Self {
+            width,
+            int_width,
+            frac_width,
+            product: Value::zeroes(width),
+            update: None,
+            queue: VecDeque::from(vec![None, None]),
+        }
+    }
+
+    pub fn new(params: &ir::Binding) -> Self {
+        let width = get_param(params, "WIDTH")
+            .expect("WIDTH parameter missing for fp mult");
+        let int_width = get_param(params, "INT_WIDTH")
+            .expect("INT_WIDTH parameter missing for fp mult");
+        let frac_width = get_param(params, "FRAC_WIDTH")
+            .expect("FRAC_WIDTH parameter missing for fp mult");
+
+        Self::from_constants(width, int_width, frac_width)
+    }
+}
+
+impl<const SIGNED: bool> Primitive for StdFpMultPipe<SIGNED> {
+    fn do_tick(&mut self) -> Vec<(ir::Id, Value)> {
+        let out = self.queue.pop_back();
+        //push update to the front
+        self.queue.push_front(self.update.take());
+        //assert queue still has length 2
+        assert_eq!(
+            self.queue.len(),
+            2,
+            "std_mult_pipe's internal queue has length {} != 2",
+            self.queue.len()
+        );
+        if let Some(Some(out)) = out {
+            vec![
+                (ir::Id::from("out"), out),
+                (ir::Id::from("done"), Value::bit_high()),
+            ]
+        } else {
+            vec![
+                (ir::Id::from("out"), Value::zeroes(self.width)),
+                (ir::Id::from("done"), Value::bit_low()),
+            ]
+        }
+    }
+
+    fn is_comb(&self) -> bool {
+        false
+    }
+
+    fn validate(&self, inputs: &[(ir::Id, &Value)]) {
+        validate![inputs;
+            left: self.width,
+            right: self.width,
+            go: 1
+        ];
+    }
+
+    fn execute(&mut self, inputs: &[(ir::Id, &Value)]) -> Vec<(ir::Id, Value)> {
+        let left = get_input_unwrap(inputs, "left");
+        let right = get_input_unwrap(inputs, "right");
+        let go = get_input_unwrap(inputs, "go");
+
+        if go.as_bool() {
+            let backing_val = if SIGNED {
```

Comment:
This conditional can be simplified a bit to hoist the Value stuff out.

Reply:
I'm not sure what you mean in this context? Could you elaborate?

Reply:
Maybe something like this?

```rs
Value::from(
  if SIGNED {
    left.as_i64().wrapping_mul(right.as_i64())
  } else {
    left.as_u64().wrapping_mul(right.as_u64())
  },
  2 * self.width,
)
```

Reply:
No dice, I'm afraid. From is generic over numeric types so the invocation would be monomorphized here and that means it wouldn't type check since the if-arms have different return types.

Reply:
The thought occurs that maybe we should rename the from method since it's not actually using the trait...

Reply:
Okay, monomorphize issue makes sense. Still super hyped about the const generic stuff.

Reply:
I do want to use them everywhere now lol, though I don't fully understand what happens when you don't specify what their value is given that you can create them without the turbofish syntax. Does it just use the `default()` for that type?

Reply:
forget the above, that doesn't appear to be true. I must've had an error elsewhere that prevented the type inference from telling me that when I was first messing around